internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -0,0 +1,671 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	// agentServiceConfigName is the one and only name for an AgentServiceConfig
+	// supported in the cluster. Any others will be ignored.
+	agentServiceConfigName = "agent"
+	// filesystemPVCName is the name of the PVC created for assisted-service's filesystem.
+	filesystemPVCName = "agent-filesystem"
+	// databasePVCName is the name of the PVC created for postgresql.
+	databasePVCName = "agent-database"
+
+	name                                = "assisted-service"
+	databaseName                        = "postgres"
+	databaseSecretName                  = databaseName
+	databasePort                  int32 = 5432
+	servicePort                   int32 = 8090
+	assistedServiceDeploymentName       = "assisted-service"
+
+	// assistedServiceContainerName is the Name property of the assisted-service container
+	assistedServiceContainerName string = "assisted-service"
+	// databaseContainerName is the Name property of the postgres container
+	databaseContainerName string = databaseName
+)
+
+// AgentServiceConfigReconciler reconciles a AgentServiceConfig object
+type AgentServiceConfigReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+	// Namespace the operator is running in
+	Namespace string
+}
+
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=agentserviceconfigs/finalizers,verbs=update
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=secrets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=route.openshift.io,resources=routes,verbs=get;list;watch;create;update;patch;delete
+
+func (r *AgentServiceConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	_ = r.Log.WithValues("agentserviceconfig", req.NamespacedName)
+
+	instance := &adiiov1alpha1.AgentServiceConfig{}
+
+	// We only support one AgentServiceConfig per cluster, and it must be called "agent". This prevents installing
+	// AgentService more than once in the cluster.
+	if req.NamespacedName.Name != agentServiceConfigName {
+		r.Log.Info(fmt.Sprintf("Invalid name (%s). Only one AgentServiceConfig supported per cluster and must be named '%s'", req.NamespacedName.Name, agentServiceConfigName))
```

Comment:
Probably worth firing an event for the resource also saying it's being ignored. That would show up in `oc describe` and the console. This could be a follow-up improvement if you want.

Reply:
![1617738473](https://user-images.githubusercontent.com/569754/113769958-9371f000-96ef-11eb-81a7-41071825ae82.png)